### PR TITLE
types: use readonly array / collection types for user input

### PR DESCRIPTION
**Pull request by almeidx**
**Please describe the changes this PR makes and why it should be merged:**
Added an ESLint rule to enforce the use of readonly Array and Collection types in function/method parameters and interface properties

**Status and versioning classification:**

### CODE REVIEW

File: packages/discord.js/typings/index.d.ts [hunk 1]
```diff
@@ -1008,7 +1016,7 @@ export class Client<Ready extends boolean = boolean> extends BaseClient {
   ): this;
   public on<Event extends string | symbol>(
     event: Exclude<Event, keyof ClientEvents>,
-    listener: (...args: any[]) => Awaitable<void>,
+    listener: (...args: readonly any[]) => Awaitable<void>,
```
Comment (vladfrangu) [inline packages/discord.js/typings/index.d.ts:1019]: Can you check this one specific change won't break anything? It also diverts from the types in node and AEE

**Comment by Jiralite** [comment]
Is there one for `const` type parameters?

File: packages/discord.js/typings/index.d.ts (hunk 1 above)
Comment (almeidx) [reply packages/discord.js/typings/index.d.ts:1019]: I'll just make an exception for `on` and `once` methods as to keep it the same as node's types

**Review by kyranet — APPROVED** [review]

**Review by Jiralite — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]
Hopefully this won't break user code

**Comment by vladfrangu** [comment]
Also this needs a rebase

**Comment by ThinkThoughtThunk** [comment]
This is a breaking change and should have been saved for a new major version number. This ended up breaking my codebase everywhere I was directly referencing one of the "user input" message / response types since I wasn't explicitly declaring the data objects as `readonly`.

**Pull request merged**